core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala

```diff
@@ -0,0 +1,173 @@
+import org.apache.kafka.common.{InvalidRecordException, Node, TopicPartition}
+import org.apache.kafka.common.message.AddPartitionsToTxnRequestData.{AddPartitionsToTxnTransaction, AddPartitionsToTxnTransactionCollection}
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.requests.{AddPartitionsToTxnRequest, AddPartitionsToTxnResponse}
+import org.apache.kafka.common.utils.Time
+
+import java.util.Collections
+import scala.collection.mutable
+
+object AddPartitionsToTxnManager {
+  type AppendCallback = Map[TopicPartition, Errors] => Unit
+}
+
+
+class TransactionDataAndCallbacks(val transactionData: AddPartitionsToTxnTransactionCollection,
+                                  val callbacks: mutable.Map[String, AddPartitionsToTxnManager.AppendCallback])
+
+
+class AddPartitionsToTxnManager(config: KafkaConfig, client: NetworkClient, time: Time) 
+  extends InterBrokerSendThread("AddPartitionsToTxnSenderThread-" + config.brokerId, client, config.requestTimeoutMs, time) {
+  
+  private val inflightNodes = mutable.HashSet[Node]()
+  private val nodesToTransactions = mutable.Map[Node, TransactionDataAndCallbacks]()
+  
+  def addTxnData(node: Node, transactionData: AddPartitionsToTxnTransaction, callback: AddPartitionsToTxnManager.AppendCallback): Unit = {
+    // Check if we have already (either node or individual transaction). 
+    val currentNodeAndTransactionDataOpt = nodesToTransactions.get(node)
+    currentNodeAndTransactionDataOpt match {
+      case None =>
+        nodesToTransactions.put(node,
+          new TransactionDataAndCallbacks(new AddPartitionsToTxnTransactionCollection(Collections.singletonList(transactionData).iterator()),
+            mutable.Map(transactionData.transactionalId() -> callback)))
+      case Some(currentNodeAndTransactionData) =>
+        // Check if we already have txn ID -- this should only happen in epoch bump case. If so, we should return error for old entry and remove from queue.
+        val currentTransactionData = currentNodeAndTransactionData.transactionData.find(transactionData.transactionalId)
+        if (currentTransactionData != null) {
+          if (currentTransactionData.producerEpoch() < transactionData.producerEpoch()) {
+            val topicPartitionsToError = mutable.Map[TopicPartition, Errors]()
+            currentTransactionData.topics().forEach { topic => 
+              topic.partitions().forEach { partition =>
+                topicPartitionsToError.put(new TopicPartition(topic.name(), partition), Errors.INVALID_PRODUCER_EPOCH)
+              }
+            }
+            val oldCallback = currentNodeAndTransactionData.callbacks(transactionData.transactionalId())
+            currentNodeAndTransactionData.transactionData.remove(transactionData)
+            oldCallback(topicPartitionsToError.toMap)
+          } else {
+            // We should never see a request on the same epoch since we haven't finished handling the one in queue
```

Comment:
I thought about that, but I was concerned about blocking on a single produce request too long. I though maybe the producer's retry mechanism would be enough to handle this.

Reply:
if 1st request timeout, the second one cannot hit this stage unless 1st one is done due to connection muted?

Reply:
Yi is correct. I also think if we hit timeout, it is the epoch bump case I mentioned before.

Reply:
Ah. I guess I was thinking of the client restarting. So you are saying in the case where the request times out before we send the request here, we can hit the error?

Reply:
Yes, that's my understanding based on my reading of the code -- once a connection to a broker has at least one timed-out in-flight request, it's disconnected and all in-flight requests get a timeout error (which can be then retried by the producer on a new connection).  https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L831

It's also not guaranteed that the request that comes later here is the retry (i.e. the assumption that we can just fail the previous request because it must've timed out anyway is not true).  Should be unlikely with default settings, so we can probably just return some retriable error and document the caveat.  Alternatively (I would actually prefer that), we could just support multiple requests per transactional.id which would eliminate the need to handle this case here altogether: just add all new requests to the pending "batch" and let the TC handle different cases.

Reply:
I don't think we can add multiple to the batch because of how the callbacks are currently implemented. We would need to know which response goes to which callback. I thought about this a while and one callback is the the simplest way to do it.

Reply:
Ok, looks like the protocol is not designed to support multiple requests for one transactional.id.  For same-epoch case, though, we'd get the same answer because we'd be asking the same question, so we could just keep multiple callbacks per trasnactional.id and call them all with the response, this way we don't have to guess which message is first and which is the retry.

Reply:
Hmm -- would it be better to just replace the old one? I don't think we'd expect two responses. And we definitely don't want to write twice.

Reply:
I assume the duplicate and out of order check happens in producer id cache and before this stage. If that's true, in general, the next batches with the same epoch won't hit this stage and should be fenced with out of order retry error to client.

There can be a special case where this is the very first batch write from new epoch so the producer id cache cannot block any batches into this stage.

Reply:
Duplicate check + store has to be atomic (cannot really discard a new request as a duplicate until the previous request succeeds, not can let it go through until the previous request fails), so it needs to happen under a lock.  The purpose of this stage is to not let a request go into log if the transaction is not there, so it got to be either between the check and the store or before the check and the store, hopefully, it's the latter, because otherwise we'd have a long lock around inter-broker RPC.

BTW, there should be no out-of-order errors or fenced errors during normal retry processing -- the first try should go through and others would be bounced with "duplicate" error which is effectively a success.  This way all tries would be effectively successful and the intermittent error would be transparently handled by Kafka without bubbling up to the application.

Reply:
> I assume the duplicate and out of order check happens in producer id cache and before this stage.

It does not happen before this stage.

But Artem is right. If we do process the callback twice, I suppose the second one will see it as a duplicate and return fine. I think my concern is that we return a timeout response + the callback response, but I guess this sort of thing can happen normally? 

I guess my question is whether we still want to return both callbacks with the response to addpartitionstotxn, or if it is just sufficient to ignore the first one.